### PR TITLE
feat(connector): implement Authorize for affirm

### PR DESCRIPTION
**Pull request by shuklatushar226**
## Summary

Implement **Authorize** flow for the **Affirm** connector.

This implementation was generated and validated by **GRACE** (automated connector integration pipeline).

Scope: Affirm one-time PayLater (BNPL redirect) payments only — **Authorize** plus supporting **PSync, Capture, Refund, RSync, Void**. No mandates, MIT, recurring, or tokenization.

## Changes

- Added Authorize (and PSync/Capture/Refund/RSync/Void) support to `affirm.rs` (`create_all_prerequisites!` macro, `macro_connector_implementation!`)
- Added request/response types and `TryFrom` implementations in `affirm/transformers.rs`
- HTTP Basic auth: `base64(public_key:private_key)`, credentials read at runtime from request context (no hardcoded secrets)
- Connector registration: `connectors.rs`, `types.rs`, `default_implementations.rs`
- Domain types: `domain_types/src/{router_data.rs,connector_types.rs,types.rs}`
- Proto contract: `grpc-api-types/proto/payment.proto`
- Field probe auth: `field-probe/src/auth.rs`
- Base URLs (no secrets): `config/{development,sandbox,production}.toml`

## Files Modified

- `crates/integrations/connector-integration/src/connectors/affirm.rs`
- `crates/integrations/connector-integration/src/connectors/affirm/transformers.rs`
- `crates/integrations/connector-integration/src/connectors.rs`
- `crates/integrations/connector-integration/src/types.rs`
- `crates/integrations/connector-integration/src/default_implementations.rs`
- `crates/types-traits/domain_types/src/router_data.rs`
- `crates/types-traits/domain_types/src/connector_types.rs`
- `crates/types-traits/domain_types/src/types.rs`
- `crates/types-traits/grpc-api-types/proto/payment.proto`
- `crates/internal/field-probe/src/auth.rs`
- `config/development.toml`, `config/sandbox.toml`, `config/production.toml`

## Validation Checklist

- [x] `cargo build` passed with zero errors
- [x] grpcurl Authorize request reached Affirm, authenticated, and serialized correctly (golden log verified)
- [ ] grpcurl Authorize returned 2xx — not reachable in harness (requires browser-minted checkout_token; placeholder yields 404)
- [x] No credentials in committed source code
- [x] Only connector-specific and required registration/config files modified

## How to test (grpcurl)

```bash
# terminal 1: run the server (defaults to :8000)
cargo run -p grpc-server
```

Affirm uses HTTP Basic auth passed via `x-connector-config` — substitute your sandbox keys, **do not commit real keys**:

```bash
CFG='{"config":{"Affirm":{"public_key":"<AFFIRM_PUBLIC_KEY>","private_key":"<AFFIRM_PRIVATE_KEY>","base_url":"https://sandbox.affirm.com"}}}'
```

> Affirm is a **2-leg BNPL redirect**: Authorize INITIATE mints a hosted-checkout `redirect_url`; the shopper completes it in a browser (phone → OTP `123456` → plan → ACH `021000021`/`000123456789` → agree → Confirm); Affirm returns a `checkout_token`; Authorize COMPLETE (same RPC, with `redirection_response`) charges it; Capture/Refund/Void/Sync use the resulting `connector_transaction_id`. A pure-grpcurl run can't mint a real `checkout_token`, so the real 2xx outcomes below come from driving the full flow via Hyperswitch ([juspay/hyperswitch#13049](https://github.com/juspay/hyperswitch/pull/13049)).

### Authorize — INITIATE → `PaymentService/Authorize`
Mints the hosted-checkout redirect.
```bash
grpcurl -plaintext \
  -H "x-connector: affirm" \
  -H "x-connector-config: $CFG" \
  -H "x-connector-request-reference-id: affirm_auth_001" \
  -d '{
    "merchant_transaction_id": "affirm_auth_001",
    "amount": { "minor_amount": 6000, "currency": "USD" },
    "payment_method": { "affirm": {} },
    "capture_method": "AUTOMATIC",
    "return_url": "https://example.com/return",
    "customer": { "email": { "value": "buyer@example.com" }, "first_name": "John", "last_name": "Doe" },
    "address": { "billing_address": {
        "first_name": {"value":"John"}, "last_name": {"value":"Doe"},
        "line1": {"value":"123 Main St"}, "city": {"value":"San Francisco"},
        "state": {"value":"CA"}, "zip_code": {"value":"94105"}, "country_alpha2_code": "US" } }
  }' \
  localhost:8000 types.PaymentService/Authorize
```
→ `AUTHENTICATION_PENDING` + a `redirect_to_url` (`POST /api/v2/checkout/direct`). Complete it in a browser to mint the `checkout_token`.

### Authorize — COMPLETE → `PaymentService/Authorize`
Same RPC, re-invoked with the `checkout_token` from the redirect.
```bash
grpcurl -plaintext -H "x-connector: affirm" -H "x-connector-config: $CFG" \
  -d '{
    "merchant_transaction_id": "affirm_auth_001",
    "amount": { "minor_amount": 6000, "currency": "USD" },
    "payment_method": { "affirm": {} },
    "capture_method": "MANUAL",
    "redirection_response": { "params": "<CHECKOUT_TOKEN>" }
  }' \
  localhost:8000 types.PaymentService/Authorize
```
→ `POST /api/v1/transactions` → loan authorized; `connectorTransactionId` = the Affirm transaction id (e.g. `CCSL-ZSS1`).

### Capture (FULL only) → `PaymentService/Capture`
```bash
TXN=<CONNECTOR_TXN_ID>   # from COMPLETE, e.g. CCSL-ZSS1
grpcurl -plaintext -H "x-connector: affirm" -H "x-connector-config: $CFG" \
  -d '{ "connector_transaction_id": "'"$TXN"'", "amount_to_capture": { "minor_amount": 6000, "currency": "USD" } }' \
  localhost:8000 types.PaymentService/Capture
```
→ `POST /api/v1/transactions/{TXN}/capture` → `CHARGED`. Partial amount → Affirm `409 "cannot be captured for an amount unequal to authorization hold amount"` (full-capture-only).

### PSync → `PaymentService/Get`
```bash
grpcurl -plaintext -H "x-connector: affirm" -H "x-connector-config: $CFG" \
  -d '{ "connector_transaction_id": "'"$TXN"'" }' \
  localhost:8000 types.PaymentService/Get
```
→ `GET /api/v1/transactions/{TXN}?expand=events`.

### Refund (FULL) → `PaymentService/Refund`
```bash
grpcurl -plaintext -H "x-connector: affirm" -H "x-connector-config: $CFG" \
  -d '{ "connector_transaction_id": "'"$TXN"'", "payment_amount": 6000, "refund_amount": { "minor_amount": 6000, "currency": "USD" }, "reason": "customer_request" }' \
  localhost:8000 types.PaymentService/Refund
```
```json
{ "connectorRefundId": "ZXPM3BCIK278DWC2", "status": "REFUND_SUCCESS", "statusCode": 200 }
```

### RSync → `RefundService/Get`
```bash
grpcurl -plaintext -H "x-connector: affirm" -H "x-connector-config: $CFG" \
  -d '{ "connector_transaction_id": "'"$TXN"'", "connector_refund_id": "<REFUND_ID>", "payment_method_type": "AFFIRM", "refund_amount": { "minor_amount": 6000, "currency": "USD" } }' \
  localhost:8000 types.RefundService/Get
```

### Void / cancel → `PaymentService/Void`
```bash
grpcurl -plaintext -H "x-connector: affirm" -H "x-connector-config: $CFG" \
  -d '{ "connector_transaction_id": "'"$TXN"'", "cancellation_reason": "requested_by_customer", "amount": { "minor_amount": 6000, "currency": "USD" } }' \
  localhost:8000 types.PaymentService/Void
```
→ `POST /api/v1/transactions/{TXN}/void` → `cancelled`.

### Coverage (end-to-end via Hyperswitch with a real browser-completed redirect)
| Flow | Connector → Affirm | Result |
|---|---|---|
| Authorize INITIATE | `POST /api/v2/checkout/direct` | redirect minted |
| Authorize COMPLETE | `POST /api/v1/transactions` | loan authorized |
| Capture (auto & manual, full) | `POST /api/v1/transactions/{id}/capture` | `succeeded` |
| PSync | `GET /api/v1/transactions/{id}?expand=events` | synced |
| Refund | `POST /api/v1/transactions/{id}/refund` | `REFUND_SUCCESS` |
| RSync | `GET /api/v1/transactions/{id}?expand=events` | synced |
| Void | `POST /api/v1/transactions/{id}/void` | `cancelled` |
| Capture — partial | `POST /api/v1/transactions/{id}/capture` | ⛔ Affirm `409` (full-capture-only) |

Payment method: only `PayLater(AffirmRedirect)` (BNPL) — fully covered.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by XyneSpaces** [comment]
[should-fix] **New connector implementation:** The Affirm connector addition is substantial (1,300+ lines). Please verify the following before merge:

1. **Test coverage:** Are there unit tests for the transformer mappings and integration tests for the authorize flow? New connectors should include mocked HTTP responses.

2. **Error handling:** In `affirm/transformers.rs`, ensure all `ForeignTryFrom` implementations handle edge cases gracefully with proper error context via `.attach_printable()`.

3. **Timeouts and retries:** Confirm the connector configuration includes appropriate timeout values and retry policies for Affirm's API.

4. **Webhook handling:** If Affirm sends webhooks for authorization updates, ensure webhook signature verification is implemented.

5. **Currency support:** Verify that Affirm's supported currencies are documented and validated in the connector configuration.

**Comment by XyneSpaces** [comment]
[should-fix] Affirm connector status mapping may need refinement

The status mapping includes several disputed/refunded states mapping to `Unresolved`:
- `Disputed` → `Unresolved`
- `PartiallyRefunded` → `Unresolved` 
- `DisputeRefunded` → `Unresolved`

These should potentially map to more specific statuses (e.g., `Disputed` → `DisputeInitiated`, `PartiallyRefunded` → `PartialCharged` or `Pending` depending on desired UX).

Also verify the `Unknown` → `Pending` mapping is safe — if Affirm introduces new statuses, they'll appear as pending rather than failures.

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+}
+
+impl From<AffirmTransactionStatus> for AttemptStatus {
+    fn from(status: AffirmTransactionStatus) -> Self {
+        match status {
+            AffirmTransactionStatus::Authorized => Self::Authorized,
+            AffirmTransactionStatus::Captured => Self::Charged,
+            AffirmTransactionStatus::PartiallyCaptured => Self::PartialCharged,
+            AffirmTransactionStatus::Voided => Self::Voided,
+            AffirmTransactionStatus::Refunded => Self::Charged,
```
Comment (iemyashasvi) [inline crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: why is refunded mapped to charged ?

**Review by deepanshu-iiitu — DISMISSED** [review]

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs [hunk 2]
```diff
@@ -0,0 +1,828 @@
...
+    fn try_from(
+        item: ResponseRouterData<AffirmPaymentsResponse, Self>,
+    ) -> Result<Self, Self::Error> {
+        match item.response {
+            // INITIATE: send the shopper to Affirm's hosted checkout.
+            AffirmPaymentsResponse::Checkout(resp) => {
+                let redirect_url = Url::parse(&resp.redirect_url).change_context(
+                    errors::ConnectorError::ResponseHandlingFailed {
+                        context: errors::ResponseTransformationErrorContext {
+                            http_status_code: None,
```
Comment (iemyashasvi) [inline crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: dont map to `None `

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs (hunk 1 above)
Comment (shuklatushar226) [reply crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: Good catch — that arm diverged from upstream. Hyperswitch's own Affirm connector maps `Refunded | PartiallyRefunded | Disputed | DisputeRefunded => Unresolved`; a fully-refunded payment needs manual reconciliation and there is no `AttemptStatus::Refunded` (the refund itself is tracked separately via `RefundStatus`/RSync). Fixed in `37f230123` — `Refunded` now joins the `Unresolved` group, so the mapping matches upstream and the "mirrors hyperswitch" comment is now accurate.

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs (hunk 2 above)
Comment (shuklatushar226) [reply crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: Fixed in `37f230123` — now passes the real connector HTTP status (`Some(item.http_code)`) instead of `None`. The same `item.http_code` was already used for `status_code` in that response, so the error context now carries the actual status. Applied the same fix to the sibling site in the Capture response transformer for consistency.

**Review by deepanshu-iiitu — DISMISSED** [review]

**Review by iemyashasvi — DISMISSED** [review]

**Review by AmitsinghTanwar007 — DISMISSED** [review]

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs [hunk 3]
```diff
@@ -0,0 +1,843 @@
...
+    fn try_from(item: ResponseRouterData<AffirmRSyncResponse, Self>) -> Result<Self, Self::Error> {
+        let connector_refund_id = item.router_data.request.connector_refund_id.clone();
+
+        // The refund is complete when a matching refund event is present in the
+        // transaction's events array, or the transaction is marked refunded.
+        let refund_event_present = item
+            .response
+            .events
+            .as_ref()
+            .map(|events| events.iter().any(|event| event.event_type == "refund"))
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: **[S2] RSync refund-completion is ambiguous across multiple refunds.** `events.iter().any(|event| event.event_type == "refund")` matches *any* refund event on the transaction, not the one identified by `connector_refund_id`. On a second partial refund, RSync can report the first refund's event as `Success` prematurely.

`AffirmEvent.id` is available — match it against `connector_refund_id`, e.g.:
```rust
.any(|event| event.event_type == "refund" && event.id == connector_refund_id)
```
and derive `Pending`/`Success` from `amount_refunded` vs the requested refund amount.

**Review by JeevaRamu0104 — COMMENTED** [review]
Automated non-blocking findings (no S0/S1 blockers). Refund-sync robustness refinements + minor cleanups; posted inline below.

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs [hunk 4]
```diff
@@ -0,0 +1,843 @@
...
+            .unwrap_or(false);
+
+        let refund_status =
+            if refund_event_present || item.response.status == AffirmTransactionStatus::Refunded {
+                RefundStatus::Success
+            } else {
+                RefundStatus::Pending
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: **[S2] RSync has no `RefundStatus::Failure` path.** A declined/failed refund falls through to `Pending` and can never resolve, leaving the refund stuck indefinitely. Add a branch that maps declined/failed refund states (or a void-of-refund) to `RefundStatus::Failure`.

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+    for RouterDataV2<Refund, RefundFlowData, RefundsData, RefundsResponseData>
+{
+    type Error = error_stack::Report<errors::ConnectorError>;
+
+    fn try_from(item: ResponseRouterData<AffirmRefundResponse, Self>) -> Result<Self, Self::Error> {
+        // A successful (HTTP 2xx) refund event indicates the refund was accepted.
+        Ok(Self {
+            response: Ok(RefundsResponseData {
+                connector_refund_id: item.response.id.clone(),
+                refund_status: RefundStatus::Success,
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: **[S3] Refund response hardcodes `Success` on any 2xx.** No inspection of the returned refund event; if Affirm ever returns 2xx for an accepted-but-async refund it would be misreported. RSync partially mitigates, but consider deriving status from the event/`amount`.

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+                splits: None,
+            }),
+            ..item.router_data.clone()
+        })
+    }
+}
+
+// ===== CAPTURE =====
+// POST /api/v1/transactions/{id}/capture  -> returns capture event { id, type, amount }
+// `amount` is sent so partial captures are honoured; omit-for-full is not assumed.
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: **[S3] Misleading comment.** This says `amount` is sent "so partial captures are honoured," but per the PR description Affirm rejects partial captures with `409` (full-capture-only). Either fix the comment to state full-capture-only, or add an explicit local guard that rejects partial-capture requests with a clear error instead of relying on Affirm's 409.

File: crates/integrations/connector-integration/src/connectors/affirm/transformers.rs [hunk 7]
```diff
@@ -0,0 +1,843 @@
...
+            )),
+        }
+    }
+}
+
+// ===== ERROR =====
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct AffirmErrorResponse {
+    pub status_code: Option<serde_json::Value>,
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/affirm/transformers.rs]: **[S3] Dead field.** `AffirmErrorResponse.status_code: Option<serde_json::Value>` is declared but never used — `ErrorResponse.status_code` is taken from `res.status_code` in `build_error_response`. Consider removing it.

File: crates/types-traits/grpc-api-types/proto/payment.proto [hunk 8]
```diff
@@ -5260,6 +5268,8 @@ message ConnectorSpecificConfig {
     TsysTransitConfig tsys_transit = 133;
     // KOUNT = 128
     KountConfig kount = 134;
+    // AFFIRM = 129
```
Comment (JeevaRamu0104) [inline crates/types-traits/grpc-api-types/proto/payment.proto]: **[S3] Confusing comment.** `// AFFIRM = 129` sits beside `affirm = 135`; `129` is already `payconex`'s number in this oneof. Harmless (the comment appears to track a separate connector-enum numbering), but worth aligning to avoid future misreads.

**Review by JeevaRamu0104 — APPROVED** [review]

**Review by iemyashasvi — APPROVED** [review]

**Review by AmitsinghTanwar007 — APPROVED** [review]
